### ISSUE TITLE
Docs request - additional examples/clarification for using CellMeasurer with a List

### ISSUE THREAD

**Issue by bsbechtel** [bvaughn/react-virtualized#608]
With the recent changes in the CellMeasurer API, some of the docs are a little hard to follow. Could you provide additional examples or clarification on how CellMeasurer should be used with a List component? 

For my specific situation, I'm trying to understand how to write the rowRenderer function passed into a List. The List docs show the props for the rowRenderer function as:

`Responsible for rendering a row. Signature should look like ({ index: number, key: string, style: Object, isScrolling: boolean }): React.PropTypes.node and the returned element must handle index, key and style.`

Yet the CellMeasurer docs show an example as:

```
function rowRenderer ({ index, isScrolling, key, parent, style }) {
  const source // This comes from your list data

  return (
    <CellMeasurer
      cache={cache}
      columnIndex={0}
      key={key}
      parent={parent}
      rowIndex={index}
    >
      {({ measure }) => (
        <img
          onLoad={measure}
          src={source}
        />
      )}
    </CellMeasurer>
  );
}
```

The rowRenderer function used in the grid example just above on the CellMeasurer docs page has a different set of props passed into the function. The use of ES6 destructuring makes it really hard to follow what properties we have access to in each of these different contexts. Maybe a separate docs page explaining how to use rowRenderer in each of these three contexts could be helpful? Thanks

**Comment by bvaughn** [comment]
I'm sorry to hear that you've found the docs confusing. As someone who's intimately familiar with the project, it can be difficult to write docs that are approachable to users who are less familiar.

> Could you provide additional examples or clarification on how CellMeasurer should be used with a List component?

There's an example checked into the project that you may find useful (demo [here](https://bvaughn.github.io/react-virtualized/#/components/CellMeasurer), source code [here](https://github.com/bvaughn/react-virtualized/blob/master/source/CellMeasurer/CellMeasurer.DynamicHeightList.example.js)). It shows using a `CellMeasurerCache` with a `List` to measure heights of lazy-loaded images. The docs [here](https://github.com/bvaughn/react-virtualized/blob/master/docs/CellMeasurer.md#using-cellmeasurer-with-images) also show the combination of `CellMeasurer` and `List` (albeit with the more advance `measure` callback param).

> The rowRenderer function used in the grid example just above on the CellMeasurer docs page has a different set of props passed into the function.

The `List` (and `Grid` and `Table`) actually do pass a `parent` parameter to `rowRenderer`- but it isn't mentioned too prominently in the `List` docs because it's only intended for consumption with the `CellMeasurer`. You can ignore the parameter otherwise.

> The use of ES6 destructuring makes it really hard to follow what properties we have access to in each of these different contexts. Maybe a separate docs page explaining how to use rowRenderer in each of these three contexts could be helpful?

I am happy to add a more in-depth `rowRenderer` section to the `List` docs [here](https://github.com/bvaughn/react-virtualized/blob/master/docs/List.md#rowrenderer). I would also encourage you to feel free to contribute improvements to the docs yourself if you can think of additional ways to improve them.

**Comment by rclai** [comment]
Um, what if you don't have any images on your `List`? What do you do with the `measure` function?

**Comment by bvaughn** [comment]
Don't use it. That's showing how you could use a render callback, which has access to the `measure` function, to handle async measurements. If you don't need them (b'c your content is text, for example) then don't use that prop at all.

**Comment by rclai** [comment]
I'm confused, so if that's the case does my content still need to be rendered in a child function? I'm getting a really weird flicker when scrolling down.

**Comment by bvaughn** [comment]
Did you look at the docs and examples? I show an example of using `CellMeasurer` with plain text content that doesn't use the `measure` param.

If you're having troubles with the component, I suggest asking in Slack. I prefer not to use Github issues for discussion. 😄

**Comment by rclai** [comment]
You don't have an example of a `CellMeasurer` on a `List` that doesn't use `measure`. [This example](https://github.com/bvaughn/react-virtualized/blob/master/source/CellMeasurer/CellMeasurer.DynamicHeightList.example.js) is the only one I'm seeing. The other ones use `Grid` and `Table`.

**Comment by bvaughn** [comment]
If you're having troubles with the component, I suggest asking in Slack.

**Comment by bsbechtel** [comment]
@bvaughn Thanks for updating the docs. It helped immensely with my problem. I ended up looking into the source code and using the cache clear functions some because my infinite loader has different types of data going into it. If I get a chance to do a PR and expand on how to use the cache, I'd be happy to do so. It might take a little time to get it done though.

**Comment by bvaughn** [comment]
Thanks @bsbechtel! No hurry but docs PRs are always welcome. 😄

**Comment by batamire** [comment]
I'm trying to implement [InfiniteLoader](https://github.com/bvaughn/react-virtualized/blob/master/source/InfiniteLoader/InfiniteLoader.example.js) for a messaging system with text of relative height. I'm having a hard time understanding where to put `<CellMeasurer>` in `rowRenderer`... How should I call `measure()` on `STATUS_LOADED`? If I go without measure(), text overlaps once messages get loaded.

Docs are really hard for someone new with React :)

**Comment by bvaughn** [comment]
This [minimal Twitter app](https://tweets.now.sh/) shows a complete integration of `AutoSizer` + `List` + `CellMeasurer`. It's source code is [here](https://github.com/bvaughn/tweets). The specific component that ties things together is [here](https://github.com/bvaughn/tweets/blob/master/src/components/TweetList.js). Check that out for a reference, @batamire. 😄
